### PR TITLE
bug: ensure CompletableFuture can be used on the manifold interceptor chain

### PR DESCRIPTION
**Pull request by mping-exo**
@jeromer caught a bug where we were using `auspex` for timeouts, and the interceptor chain was not waiting for the CF to terminate, because `CompletableFuture` does ǹot satisfy `AsyncContext`.

This will enforce the minimum Java version to 1.8, because of the `import` (CF's are only available from java 1.8).

### CODE REVIEW

**Comment by mping-exo** [comment]
I wonder if we should do the same for java's `Future` and clojure promises, because manifold supports them too

**Comment by jeromer** [comment]
> This will enforce the minimum Java version to 1.8, because of the import (CF's are only available from java 1.8).

@pyr would this requirement be acceptable to you ?

**Comment by pyr** [comment]
Yes, we have no JDK <8 running anywhere (all Clojure code runs on JDK11 iirc)

**Comment by mpenet** [comment]
If you require interceptor.auspex (cf/cs support) in your code then cf extends asyncontext. It's not something I would mix in interceptor.manifold.

**Comment by mping-exo** [comment]
We already rely on manifold for deferred, it seems to me we should allow whatever manifold allows.

Since interceptor already requires `auspex` as a dependency, why not make this behaviour the default? It seems unexpected that I have to require auspex to make interceptors work with CF, since manifold doesn't make you require anything?

**Comment by mpenet** [comment]
> We already rely on manifold for deferred, it seems to me we should allow whatever manifold allows.

That isn't a valid reason imho, the manifold interface is supposed to work with deferreds and return deferreds and nothing else. You're free to call d/->deferred on the input to coerce the input.

> Since interceptor already requires auspex as a dependency, why not make this behaviour the default? It seems unexpected that I have to require auspex to make interceptors work with CF, since manifold doesn't make you require anything?

We do not include auspex or manifold as a dependency, it's the responsibility of the user to do so if they require one or the other, we don't assume auspex is present if manifold is or vis-versa.

The main issue with that patch is not that really; it's that it will make all CompletableFuture be composed through manifold chain if it is present, and manifold required last, or just break if interceptor.auspex is required last because all this relies on the order of extend-protocol since they extend the same Class. In the former case it will cause overhead you want to avoid when using CompletableFuture compared to manifold usually, and in the later it will just break at runtime. So this will cause inconsistencies that shouldn't be.Throw in some AOT in the mix and it's a nightmare to debug.

What we could do is assume if it's the jvm via reader conditional automatically extend for CF in the main namespace, but that would have to be done without auspex present likely since we don't want to impose it as a dependency, so imho not worth it really. For the manifold+auspex usage as I mentioned before, just require both nses if you'd like, you just have to do it once, not necessary everywhere you use it and that's safe since they work on different types.

In short I really do not consider this issue a bug at all, it's working as intended.

**Comment by mping-exo** [comment]
You're correct, I missed it was a dev dependency. I thought it was a regular dependency. The order of requires can also be an issue like you mention, I didn't catch that. I will resort to `d/->deferred`; I feel relying on a require to activate a feature breaks the principle of least surprise, although its somewhat common in clojure.

Thanks

**Comment by mpenet** [comment]
I wouldn't close the door to making cf a default and not even force a require to get support. It just require a bit of effort to do it and it might be a good thing. We can imagine the same for cljs promises & so on, as long as it's available on the host runtime we want to support. But personally I have too many yaks half-shaven right now to tackle this.

**Comment by mping-exo** [comment]
Same, I may address this later in another PR, specially if it comes up as a surprise again.
